pkg/engine/supervisor.go

```diff
@@ -70,13 +70,42 @@ func (e *Engine) worker(n int) {
 
 			ch := make(chan int)
 			e.addSignal(tsk.ID, ch)
-
 			go func() {
 				select {
 				case <-ch:
 					e.deleteSignal(tsk.ID)
 					cancel()
 				case <-ctx.Done():
+					// Timeout reached,
+					// 1. Remove timeouted tasks from the queue -- It is going to be removed anyway, since we are popping from the queue.
+					// 2. Stop timeouted task execution (e.g. stopping the docker container)
+					e.deleteSignal(tsk.ID)
+					// 3.  Update the state of the task in the database
+					logging.S().Infow("Timeout reached for task: ", "task_id", tsk.ID)
+					// get the current state of the task from DB.
+					tsk, err := e.store.Get(tsk.ID)
+					if err != nil {
+						return
+					}
+					// if task is still in processing state, mark it as canceled in DB
```

Comment:
I also thought  about it, but I guess why it is not implemented that way firstly is that it doesn't mean that task is finished when the container is stopped, also there can be multiple containers for one task, so that's why the update is done from the same 'hierarchy' level as task starting is.